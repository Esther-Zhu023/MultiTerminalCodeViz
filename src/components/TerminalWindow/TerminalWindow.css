.react-resizable {
  position: relative;
  /* background-clip: padding-box; */ /* Optional: Might conflict with terminal styling */
}

.react-resizable-handle {
  position: absolute;
  width: 20px;
  height: 20px;
  background-repeat: no-repeat;
  background-origin: content-box;
  box-sizing: border-box;
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA2IDYiIHN0eWxlPSJiYWNrZ3JvdW5kLWNvbG9yOndoaXRlMDAiIHg9IjAiIHk9IjAiIHdpZHRoPSI2cHgiIGhlaWdodD0iNnB4Ij48ZyBvcGFjaXR5PSIwLjMwMiI+PHBhdGggZD0iTSA2IDYgTCAwIDYgTCAwIDQuOCBMIDQuOCA0LjggTCA0LjggMCBMIDYgMCBMIDYgNiBaIiBmaWxsPSIjMDAwMDAwIi8+PC9nPjwvc3ZnPg==');
  background-position: bottom right;
  padding: 0 3px 3px 0;
}

.react-resizable-handle-sw {
  bottom: 0;
  left: 0;
  cursor: sw-resize;
  transform: rotate(90deg);
}

.react-resizable-handle-se {
  bottom: 0;
  right: 0;
  cursor: se-resize;
}

.react-resizable-handle-nw {
  top: 0;
  left: 0;
  cursor: nw-resize;
  transform: rotate(180deg);
}

.react-resizable-handle-ne {
  top: 0;
  right: 0;
  cursor: ne-resize;
  transform: rotate(270deg);
}

.react-resizable-handle-w,
.react-resizable-handle-e {
  top: 50%;
  margin-top: -10px;
  cursor: ew-resize;
}

.react-resizable-handle-w {
  left: 0;
  transform: rotate(135deg);
}

.react-resizable-handle-e {
  right: 0;
  transform: rotate(315deg);
}

.react-resizable-handle-n,
.react-resizable-handle-s {
  left: 50%;
  margin-left: -10px;
  cursor: ns-resize;
}

.react-resizable-handle-n {
  top: 0;
  transform: rotate(225deg);
}

.react-resizable-handle-s {
  bottom: 0;
  transform: rotate(45deg);
}

/* Ensure the resizable box itself uses absolute positioning if it's inside a Draggable that will be providing x/y */
.react-resizable-box.absolute {
    position: absolute;
}

/* Hide scrollbars */
.terminal-content {
    /* Hide scrollbar for Chrome, Safari and Opera */
    -webkit-scrollbar: none;
    -ms-overflow-style: none;  /* IE and Edge */
    scrollbar-width: none;  /* Firefox */
}

.terminal-content::-webkit-scrollbar {
    display: none;
} 