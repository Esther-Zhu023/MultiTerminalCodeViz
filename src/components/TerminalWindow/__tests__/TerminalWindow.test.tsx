import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { TerminalWindow } from '../TerminalWindow';

describe('<TerminalWindow />', () => {
  it('should match snapshot for static UI', () => {
    const { container } = render(
      <TerminalWindow 
        id="snapshot-test-window"
        title="Test Terminal"
        initialPosition={{ x: 10, y: 20 }}
        initialSize={{ width: 500, height: 300 }}
      />
    );
    // For snapshot testing with react-draggable and react-resizable, 
    // it's often best to assert against the direct child of the ResizableBox
    // or a stable part of your component structure, as the wrapper elements
    // from these libraries can add dynamic attributes (like transform styles)
    // that make snapshots brittle.
    // Here, we target the main 'terminal-window' div directly.
    const terminalWindowElement = container.querySelector('.terminal-window');
    expect(terminalWindowElement).toMatchSnapshot();
  });

  // 3.4 Clamping tests for resize (verifying props passed to ResizableBox)
  it('should pass correct min/max constraints to ResizableBox', () => {
    const { container } = render(
      <TerminalWindow 
        id="constraints-test"
        initialSize={{ width: 300, height: 200 }}
      />
    );

    // ResizableBox is the direct child of Draggable. 
    // We inspect its props indirectly by looking at what our TerminalWindow renders.
    // This test is more about ensuring our component configures ResizableBox correctly.
    // Note: Directly accessing props of a rendered component is often an anti-pattern
    // in RTL, but for HOCs/wrapper components, ensuring they are configured correctly is key.
    // A more robust way might involve a more complex setup or mocking ResizableBox to spy on its props.
    
    // For now, we know ResizableBox applies these as inline styles or attributes eventually,
    // but testing the direct output HTML for min/max constraints is not straightforward with ResizableBox
    // as it uses JavaScript calculations. We are trusting that if we pass the props correctly,
    // react-resizable will do its job.

    // What we can do is render with specific initial sizes and ensure they are passed.
    const resizableBoxElement = container.firstChild; // This should be the Draggable component
    const actualResizableBox = resizableBoxElement?.firstChild; // This should be the ResizableBox

    // This is a limitation. We can't easily assert the props of ResizableBox post-render with RTL.
    // The snapshot test above will cover the initial dimensions.
    // The best we can do here is acknowledge the constraints are hardcoded in TerminalWindow.
    // If we were to make min/maxConstraints dynamic props of TerminalWindow, we could test those.
    // For now, we will assume the hardcoded [200,100] and [1200,800] are passed.
    // This test serves more as a placeholder for this known testing challenge.

    // A simple structural check to ensure ResizableBox is there:
    // The className "react-resizable" is added by ResizableBox to its wrapper div.
    expect(actualResizableBox).toHaveClass('react-resizable');

    // We can also check that our initial size is applied.
    // The ResizableBox wrapper div gets inline styles for width and height.
    expect(actualResizableBox).toHaveStyle('width: 300px;');
    expect(actualResizableBox).toHaveStyle('height: 200px;');
  });

  it('should apply default dimensions if initialSize is not provided', () => {
    const { container } = render(<TerminalWindow id="default-size-test" />);
    const resizableBoxElement = container.firstChild?.firstChild;
    expect(resizableBoxElement).toHaveStyle('width: 600px;');
    expect(resizableBoxElement).toHaveStyle('height: 400px;');
  });

  it('should apply partial default dimensions if initialSize is partially provided', () => {
    const { container: containerW } = render(<TerminalWindow id="partial-w" initialSize={{ width: 350 }} />);
    const resizableBoxW = containerW.firstChild?.firstChild;
    expect(resizableBoxW).toHaveStyle('width: 350px;');
    expect(resizableBoxW).toHaveStyle('height: 400px;'); // Default height

    const { container: containerH } = render(<TerminalWindow id="partial-h" initialSize={{ height: 250 }} />);
    const resizableBoxH = containerH.firstChild?.firstChild;
    expect(resizableBoxH).toHaveStyle('width: 600px;'); // Default width
    expect(resizableBoxH).toHaveStyle('height: 250px;');
  });
}); 