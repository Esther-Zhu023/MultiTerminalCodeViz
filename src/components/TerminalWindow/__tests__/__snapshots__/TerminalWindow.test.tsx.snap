// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TerminalWindow /> > should match snapshot for static UI 1`] = `
<div
  class="terminal-window bg-black border-black rounded overflow-hidden w-full h-full flex flex-col subpixel-antialiased"
  id="terminal-snapshot-test-window"
>
  <div
    class="terminal-title-bar bg-gray-400 border-b border-gray-500 h-6 flex items-center text-center text-black select-none cursor-move"
  >
    <div
      class="flex ml-2 items-center space-x-2 no-drag"
    >
      <div
        class="border-red-900 bg-red-500 shadow-inner rounded-full w-3 h-3 cursor-pointer hover:bg-red-600 transition-colors"
        title="Close terminal"
      />
      <div
        class="border-yellow-900 bg-yellow-500 shadow-inner rounded-full w-3 h-3 cursor-pointer"
      />
      <div
        class="border-green-900 bg-green-500 shadow-inner rounded-full w-3 h-3 cursor-pointer"
      />
    </div>
    <div
      class="mx-auto pr-16"
    >
      <p
        class="text-center text-sm"
      >
        Test Terminal
      </p>
    </div>
  </div>
  <div
    class="terminal-content flex-grow pl-2 pr-2 pt-1 h-auto text-green-200 font-mono text-xs bg-black overflow-y-auto select-text text-left"
  >
    <pre>
      <div
        class="pb-2 text-purple-400 font-mono text-xs leading-tight"
      >
        <div>
          ██╗   ██╗   ██╗██╗██████╗ ███████╗   ███╗   ███╗ ██████╗ ██████╗ ███████╗
        </div>
        <div>
          ██║   ██║   ██║██║██╔══██╗██╔════╝   ████╗ ████║██╔═══██╗██╔══██╗██╔════╝
        </div>
        <div>
          ██║   ██║   ██║██║██████╔╝█████╗     ██╔████╔██║██║   ██║██████╔╝█████╗  
        </div>
        <div>
          ██║   ╚██╗ ██╔╝██║██╔══██╗██╔══╝     ██║╚██╔╝██║██║   ██║██╔══██╗██╔══╝  
        </div>
        <div>
          ██║    ╚████╔╝ ██║██████╔╝███████╗   ██║ ╚═╝ ██║╚██████╔╝██║  ██║███████╗
        </div>
        <div>
          ╚═╝     ╚═══╝  ╚═╝╚═════╝ ╚══════╝   ╚═╝     ╚═╝ ╚═════╝ ╚═╝  ╚═╝╚══════╝
        </div>
        <div />
        <div>
          ████████╗██╗  ██╗ █████╗ ███╗   ██╗   ██╗   ██╗ ██████╗ ██╗   ██╗
        </div>
        <div>
          ╚══██╔══╝██║  ██║██╔══██╗████╗  ██║   ╚██╗ ██╔╝██╔═══██╗██║   ██║
        </div>
        <div>
             ██║   ███████║███████║██╔██╗ ██║    ╚████╔╝ ██║   ██║██║   ██║
        </div>
        <div>
             ██║   ██╔══██║██╔══██║██║╚██╗██║     ╚██╔╝  ██║   ██║██║   ██║
        </div>
        <div>
             ██║   ██║  ██║██║  ██║██║ ╚████║      ██║   ╚██████╔╝╚██████╔╝
        </div>
        <div>
             ╚═╝   ╚═╝  ╚═╝╚═╝  ╚═╝╚═╝  ╚═══╝      ╚═╝    ╚═════╝  ╚═════╝ 
        </div>
      </div>
    </pre>
    <p
      class="pb-1 text-gray-400"
    >
      Last login: 
      Wed, Jun 12, 20:41:30
       on ttys002
    </p>
    <div
      class="pb-1"
    >
      <span
        class="text-blue-400"
      >
        user@localhost
      </span>
      <span
        class="text-green-200"
      >
        :
      </span>
      <span
        class="text-blue-400"
      >
        ~
      </span>
      <span
        class="text-green-200"
      >
        $ 
      </span>
      <span
        class="text-gray-400"
      >
        npm run dev
      </span>
    </div>
  </div>
</div>
`;
